File: app/src/main/java/com/example/schoolairdroprefactoredition/utils/RSACoder.java
package com.example.schoolairdroprefactoredition.utils;

import android.util.Base64;

import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.SignatureException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;

import javax.crypto.Cipher;

/**
 * RSA加密解密
 */
public class RSACoder {
    public static final String SIGNATURE_ALGORITHM_MD5withRSA = "MD5withRSA";
    public static final String SIGNATURE_ALGORITHM_SHA1withRSA = "SHA1withRSA";
    public static final String PUBLIC_KEY = "RSAPublicKey";
    public static final String PRIVATE_KEY = "RSAPrivateKey";
    private static final String KEY_ALGORITHM = "RSA";

    /**
     * 签名
     *
     * @param data
     * @param privateKey
     * @return
     * @throws Exception
     */
    public static byte[] sign(byte[] data, byte[] privateKey) throws Exception {
        return sign(data, privateKey, SIGNATURE_ALGORITHM_SHA1withRSA);
    }

    /**
     * 签名
     *
     * @param data
     * @param privateKey
     * @param signatureAlgorithm
     * @return
     * @throws Exception
     */
    public static byte[] sign(byte[] data, byte[] privateKey, String signatureAlgorithm) {
        PKCS8EncodedKeySpec pkcs8KeySpec = new PKCS8EncodedKeySpec(privateKey);

        KeyFactory keyFactory = null;
        try {
            keyFactory = KeyFactory.getInstance(KEY_ALGORITHM);
            PrivateKey priKey = keyFactory.generatePrivate(pkcs8KeySpec);

            Signature signature = Signature.getInstance(signatureAlgorithm);
            signature.initSign(priKey);
            signature.update(data);
            return signature.sign();
        } catch (NoSuchAlgorithmException | InvalidKeySpecException | InvalidKeyException | SignatureException e) {
            e.printStackTrace();
        }
        return "".getBytes();
    }

    /**
     * 验证
     *
     * @param data
     * @param publicKey
     * @param sign
     * @return
     * @throws Exception
     */
    public static boolean verify(byte[] data, byte[] publicKey, byte[] sign)
            throws Exception {
        return verify(data, publicKey, sign, SIGNATURE_ALGORITHM_SHA1withRSA);
    }

    /**
     * 验证
     *
     * @param data
     * @param publicKey
     * @param sign
     * @param signatureAlgorithm
     * @return
     * @throws Exception
     */
    public static boolean verify(byte[] data, byte[] publicKey, byte[] sign,
                                 String signatureAlgorithm) throws Exception {
        X509EncodedKeySpec keySpec = new X509EncodedKeySpec(publicKey);

        KeyFactory keyFactory = KeyFactory.getInstance(KEY_ALGORITHM);

        PublicKey pubKey = keyFactory.generatePublic(keySpec);

        Signature signature = Signature.getInstance(signatureAlgorithm);
        signature.initVerify(pubKey);
        signature.update(data);

        return signature.verify(sign);
    }


    /***************************************公钥加密*************************************/

    public static String encryptWithPublicKey(String publicK, String textToEncrypt) {
        String encoded = "";
        byte[] encrypted;
        try {
            PublicKey pubKey = getFromString(publicK);
            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1PADDING"); //or try with "RSA"
            cipher.init(Cipher.ENCRYPT_MODE, pubKey);
            encrypted = cipher.doFinal(textToEncrypt.getBytes(StandardCharsets.UTF_8));
            encoded = Base64.encodeToString(encrypted, Base64.DEFAULT);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return encoded;
    }

    /**************************************************************************************/


    public static PublicKey getFromString(String keystr) {
        try {
            String pubKeyDER = keystr.replace("-----BEGIN PUBLIC KEY-----", "");
            pubKeyDER = pubKeyDER.replace("-----END PUBLIC KEY-----", "");
            pubKeyDER = pubKeyDER.replace("\n", "");

            byte[] encoded = Base64.decode(pubKeyDER, Base64.NO_WRAP);
            X509EncodedKeySpec keySpec = new X509EncodedKeySpec(encoded);
            KeyFactory kf = KeyFactory.getInstance(KEY_ALGORITHM);

            return kf.generatePublic(keySpec);
        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
            e.printStackTrace();
        }
        return null;
    }

}

File: app/src/main/java/com/example/schoolairdroprefactoredition/utils/RSACoder.kt
package com.example.schoolairdroprefactoredition.utils

import android.util.Base64
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.*
import javax.crypto.Cipher

/**
 * RSA加密解密
 */
object RSACoder {

    const val PUBLIC_KEY = "RSAPublicKey"

    const val PRIVATE_KEY = "RSAPrivateKey"

    private const val KEY_ALGORITHM = "RSA"

    private const val KEY_LENGTH = 2048

    /**
     * 使用公钥加密内容
     *
     * @param publicKey       公钥
     * @param textToEncrypt 需要加密的内容
     * @return 使用公钥加密后的内容
     */
    fun encryptWithPublicKey(publicKey: String, textToEncrypt: String): String? {
        try {
            val pubKey = getPublicKeyFromString(publicKey)
            val cipher = Cipher.getInstance("RSA/ECB/PKCS1PADDING")
            cipher.init(Cipher.ENCRYPT_MODE, pubKey)
            val encrypted = cipher.doFinal(textToEncrypt.toByteArray(StandardCharsets.UTF_8))
            return Base64.encodeToString(encrypted, Base64.DEFAULT)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 私钥解密
     *
     * @param textToDecrypt       需要解密的内容
     * @param privateKey 私钥
     * @return 私钥解密后的内容
     */
    fun decryptWithPrivateKey(privateKey: String, textToDecrypt: String): String? {
        try {
            val privateK = getPrivateKeyFromString(privateKey)
            val cipher = Cipher.getInstance("RSA/ECB/PKCS1PADDING")
            cipher.init(Cipher.DECRYPT_MODE, privateK)
            val decrypted = cipher.doFinal(textToDecrypt.toByteArray(StandardCharsets.UTF_8))
            return Base64.encodeToString(decrypted, Base64.DEFAULT)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 从row公钥获取被修剪的公钥
     */
    private fun getPublicKeyFromString(public: String): PublicKey? {
        try {
            // 去掉头部和尾部
            val pubKeyDER = public.replace("-----BEGIN PUBLIC KEY-----", "")
                    .replace("-----END PUBLIC KEY-----", "")
                    .replace("\n", "")

            val encoded = Base64.decode(pubKeyDER, Base64.NO_WRAP)
            val keySpec = X509EncodedKeySpec(encoded)
            val kf = KeyFactory.getInstance(KEY_ALGORITHM)
            return kf.generatePublic(keySpec)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 从row私钥获取被修剪的私钥
     */
    private fun getPrivateKeyFromString(private: String): PrivateKey? {
        try {
            // 去掉头部和尾部
            val priKeyDER = private.replace("-----BEGIN PRIVATE KEY-----", "")
                    .replace("-----END PRIVATE KEY-----", "")
                    .replace("\n", "")

            val encoded = Base64.decode(priKeyDER, Base64.NO_WRAP)
            val keySpec = PKCS8EncodedKeySpec(encoded)
            val kf = KeyFactory.getInstance(KEY_ALGORITHM)
            return kf.generatePrivate(keySpec)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 生成一对密钥
     *
     * @return 生成的密钥对
     */
    fun generateKeys(): Map<String, String>? {
        val keyMap: MutableMap<String, String> = HashMap(2)
        try {
            val keyPairGen = KeyPairGenerator.getInstance(KEY_ALGORITHM).also {
                it.initialize(KEY_LENGTH)
            }
            val keyPair = keyPairGen.generateKeyPair()
            keyMap[PUBLIC_KEY] = Base64.encodeToString(keyPair.public.encoded, Base64.DEFAULT)
            keyMap[PRIVATE_KEY] = Base64.encodeToString(keyPair.private.encoded, Base64.DEFAULT)
            return keyMap
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }
}